crates/bevy_asset/src/server/mod.rs

```diff
@@ -1336,6 +1339,110 @@ impl AssetServer {
+    /// This may lead to the asset unexpectedly being dropped while you are waiting for it to
+    /// finish loading.
+    ///
+    /// When calling this method, make sure a strong handle is stored elsewhere to prevent the
+    /// asset from being dropped.
+    /// If you have access to an asset's strong [`Handle`], you should prefer to call
+    /// [`AssetServer::wait_for_asset`]
+    /// or [`wait_for_assest_untyped`](Self::wait_for_asset_untyped) to ensure the asset finishes
+    /// loading.
+    ///
+    /// # Errors
+    ///
+    /// This will return an error if the asset or any of its dependencies fail to load,
+    /// or if the asset has not been queued up to be loaded.
+    pub async fn wait_for_asset_id(
+        &self,
+        id: impl Into<UntypedAssetId>,
+    ) -> Result<(), WaitForAssetError> {
+        let id = id.into();
+        core::future::poll_fn(move |cx| {
+            let infos = self.data.infos.read();
+            let info = infos.get(id).ok_or(WaitForAssetError::NotLoaded)?;
+            match (&info.load_state, &info.rec_dep_load_state) {
+                (LoadState::Loaded, RecursiveDependencyLoadState::Loaded) => Poll::Ready(Ok(())),
+                // Return an error immediately if the asset is not in the process of loading
+                (LoadState::NotLoaded, _) => Poll::Ready(Err(WaitForAssetError::NotLoaded)),
+                // If the asset is loading, leave our waker behind
+                (LoadState::Loading, _)
+                | (_, RecursiveDependencyLoadState::Loading)
+                | (LoadState::Loaded, RecursiveDependencyLoadState::NotLoaded) => {
+                    // Check if our waker is already there
+                    let has_waker = info
+                        .waiting_tasks
+                        .iter()
+                        .any(|waker| waker.will_wake(cx.waker()));
+                    if !has_waker {
+                        drop(infos);
+                        let mut infos = self.data.infos.write();
+                        let info = infos.get_mut(id).ok_or(WaitForAssetError::NotLoaded)?;
+                        // If the load state changed while reacquiring the lock, immediately
+                        // reawaken the task
+                        let is_loading = matches!(
+                            (&info.load_state, &info.rec_dep_load_state),
+                            (LoadState::Loading, _)
+                                | (_, RecursiveDependencyLoadState::Loading)
+                                | (LoadState::Loaded, RecursiveDependencyLoadState::NotLoaded)
+                        );
+                        if !is_loading {
+                            cx.waker().wake_by_ref();
+                        } else {
+                            // Leave our waker behind
+                            info.waiting_tasks.push(cx.waker().clone());
+                        }
+                    }
+                    Poll::Pending
```

Comment:
Can we make `if !has_waker` a guard clause instead? As in `if has_waker { return Poll::Pending; } ... the unnested body`.

Reply:
I've done that _and_ moved the `poll_fn` into its own function to further reduce the indentation level.